consensus/istanbul/core/core.go

```diff
@@ -376,26 +376,28 @@ func (c *core) waitForDesiredRound(r *big.Int) {
 
 func (c *core) resetRoundState(view *istanbul.View, validatorSet istanbul.ValidatorSet, roundChange bool) {
 	// TODO(Joshua): Include desired round here.
-	if c.current != nil {
-		if roundChange {
-			c.current = newRoundState(view, validatorSet, c.current.PendingRequest(), c.current.PreparedCertificate(), c.current.ParentCommits())
-		} else {
-			lastSubject, err := c.backend.LastSubject()
-			if err != nil && c.current.Proposal() != nil && c.current.Proposal().Hash() == lastSubject.Digest && c.current.Round().Cmp(lastSubject.View.Round) == 0 {
-				// When changing sequences, if our current Commit messages match the latest block in the chain
-				// (i.e. they're for the same block hash and round), we use this sequence's commits as the ParentCommits field
-				// in the next round.
-				c.current = newRoundState(view, validatorSet, nil, istanbul.EmptyPreparedCertificate(), c.current.Commits())
-			} else {
-				headBlock := c.backend.GetCurrentHeadBlock()
-				// Otherwise, we will initialize an empty ParentCommits field with the validator set of the last proposal.
-				c.current = newRoundState(view, validatorSet, nil, istanbul.EmptyPreparedCertificate(), newMessageSet(c.backend.ParentBlockValidators(headBlock)))
-			}
-		}
-	} else {
+	if c.current == nil {
 		// When the current round is nil, we must start with the current validator set in the parent commits
 		// either `validatorSet` or `backend.Validators(lastProposal)` works here
-		c.current = newRoundState(view, validatorSet, nil, istanbul.EmptyPreparedCertificate(), newMessageSet(validatorSet))
+		c.current = newRoundState(view, validatorSet)
+	} else if roundChange {
```

Comment:
When there is a round change or sequence change, the `current` variable gets mutated.  Originally, a new one was created.

You could argue that there should be a new one for every round (it is a `roundState` instead of a `sequenceState`).  Also, I feel it may introduce more bugs that we aren't aware of with a change that is this significant.  WDYT?

Reply:
I agree that is a change that should be done/reviewed carefully. Still i think it's for the better; and if we can/want to do it, we should do it sooner (before the audit) than later. If not, it will never be a good time to clean up this code.

There's actual a reason for the change, besides cleaning up. I wanted to leave all mutation to `c.current` behind a single interface (now `RoundState`) so i could easily intercept all those mutations and trigger roundState persistence just after the mutation. Of course, you can try do the same, but then you would need to make `Core` persistence aware, and you will now 2 different places where mutations happens

 1. When you create a new RoundState (eg `c.current = ...`)
 2. When you mutate the current one (eg `c.current.SomeMutationMethod()`)

Now for the actual review of the change, a few things that i have considered:

* `c.current` is only ever used inside core, and core event handling is single threaded (as we can see in `Start()`. 
* In fact there was not lock when overwriting `c.current`. It makes sense since there's only one user, and that user uses a single thread.
* There were 3 cases to set `c.current`:
     * When it was unset (only on `Start()`)
     * When we are on `roundChange`
     * When we are at sequence change
* In all these situations, we called `newRoundState()` and decide which fields we wanted to leave with the old `c.current` information. What i did was to verify that the resulting fields after calling each `newRoundChange` on all 3 different scenarios, matches with all the resulting fields after replacing those with methods. That of course, it will be good to have a second pairs of eyes. But; i believe that's the only important consideration we should do, since having the same or a different instance of RoundState doesn't generate a change in behaviour from my review.

In general, I agree with you that we should be careful about this kind of changes. But; i don't think it should stop us from doing them, since that would deny us the possibility of refactoring for the better (if we agree it reads/follows better). And it's better to do those sooner than later.

Again, i might be mistaken; and this generates a change in behaviour that i didn't forsee. Do you have a concrete idea of how that could be happening?

Reply:
My response is inline:

> I agree that is a change that should be done/reviewed carefully. Still i think it's for the better; and if we can/want to do it, we should do it sooner (before the audit) than later. If not, it will never be a good time to clean up this code.
> 
> There's actual a reason for the change, besides cleaning up. I wanted to leave all mutation to `c.current` behind a single interface (now `RoundState`) so i could easily intercept all those mutations and trigger roundState persistence just after the mutation. Of course, you can try do the same, but then you would need to make `Core` persistence aware, and you will now 2 different places where mutations happens
> 
> 1. When you create a new RoundState (eg `c.current = ...`)
> 2. When you mutate the current one (eg `c.current.SomeMutationMethod()`)
> 

Keep in mind that other mutations could be happening to the state.  E.g. the `prepares` message set can at added to when a `prepare` message is received.  And I think that we would need to persist that mutation as well.

> Now for the actual review of the change, a few things that i have considered:
> 
> * `c.current` is only ever used inside core, and core event handling is single threaded (as we can see in `Start()`.
> * In fact there was not lock when overwriting `c.current`. It makes sense since there's only one user, and that user uses a single thread.
> * There were 3 cases to set `c.current`:
>   
>   * When it was unset (only on `Start()`)
>   * When we are on `roundChange`
>   * When we are at sequence change
> * In all these situations, we called `newRoundState()` and decide which fields we wanted to leave with the old `c.current` information. What i did was to verify that the resulting fields after calling each `newRoundChange` on all 3 different scenarios, matches with all the resulting fields after replacing those with methods. That of course, it will be good to have a second pairs of eyes. But; i believe that's the only important consideration we should do, since having the same or a different instance of RoundState doesn't generate a change in behaviour from my review.
> 
> In general, I agree with you that we should be careful about this kind of changes. But; i don't think it should stop us from doing them, since that would deny us the possibility of refactoring for the better (if we agree it reads/follows better). And it's better to do those sooner than later.
> 
> Again, i might be mistaken; and this generates a change in behaviour that i didn't forsee. Do you have a concrete idea of how that could be happening?

Thoughts on renaming it to `SequenceState`?  If there is only one instance, then seems like that's more accurate.